### PR TITLE
Feat: PowerSync encryption with SQLCipher

### PR DESCRIPTION
**Pull request by mugikhan**
## Description

This enables the use of database encryption with SQLCipher. It also includes a refactor into two new packages, `powersync_core` (Dart only) and `powersync_sqlcipher` in order to achieve this.
`powersync_core` includes all of the original `powersync` setup and functionality and exports it for the other packages to use.

 - `powersync_core` is now used by `powersync` which is a thin package that depends on `sqlite3_flutter_libs` for Flutter projects
 - `powersync_core` is also used by `powersync_sqlcipher` which is a thin package that depends on `sqlcipher_flutter_libs`.
 - `powersync_attachments_helper` also depends on `powersync_core` for its basic functionality.
 
 The reason for this refactor is so that we can make `PowerSync` compatible with both Dart and Flutter apps. It also allows us to enable encryption for users who can now install and import `powersync_sqlcipher`. Finally, it maintains the original PowerSync API for users using `powersync`.
 
## TODO

- [x] Update READMEs

### CODE REVIEW

File: melos.yaml [hunk 1]
```diff
@@ -33,7 +33,7 @@ scripts:
 
   prepare:demos:
     description: Download SQLite3 wasm for demos
-    run: dart run powersync:setup_web --no-worker
+    run: dart run powersync_core:setup_web --no-worker
```
Comment (rkistner) [inline melos.yaml:36]: Should this rather stay in the `powersync` package?

1. Existing applications may depend on having `dart run powersync:setup_web` available.
2. The downloaded WASM files are specific to plain SQLite, and powersync_sqlcipher (if we ever support that on web) would need to use different files.

File: packages/powersync_core/README.md [hunk 2]
```diff
@@ -0,0 +1,39 @@
+<!--
```
Comment (rkistner) [inline packages/powersync_core/README.md:1]: Readme should be updated.

File: packages/powersync_core/NOTICE [hunk 3]
```diff
@@ -0,0 +1 @@
+Copyright 2023 Journey Mobile, Inc.
```
Comment (rkistner) [inline packages/powersync_core/NOTICE:1]:
```suggestion
Copyright 2024 Journey Mobile, Inc.
```

File: packages/powersync_sqlcipher/NOTICE [hunk 4]
```diff
@@ -0,0 +1 @@
+Copyright 2023 Journey Mobile, Inc.
```
Comment (rkistner) [inline packages/powersync_sqlcipher/NOTICE:1]:
```suggestion
Copyright 2024 Journey Mobile, Inc.
```

File: packages/powersync_sqlcipher/lib/src/sqlcipher.dart [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+      {required super.path, required this.key, super.sqliteOptions});
+
+  final String key;
+
+  @override
+  List<String> pragmaStatements(SqliteOpenOptions options) {
+    final basePragmaStatements = super.pragmaStatements(options);
+    return [
+      // Set the encryption key as the first statement
+      "PRAGMA KEY = '$key'",
```
Comment (rkistner) [inline packages/powersync_sqlcipher/lib/src/sqlcipher.dart:20]:
```suggestion
      "PRAGMA KEY = ${quoteString(key)}",
```

This makes sure there is no SQL injection possibility if user input is directly used as the key. The `quoteString` function is part of sqlite_async.

File: melos.yaml (hunk 1 above)
Comment (mugikhan) [reply melos.yaml]: My original thought was that we could reuse it for SQLCipher but it isn't web compatible at this point so I moved it back into the `powersync` package.

File: packages/powersync_attachments_helper/example/getting_started.dart [hunk 6]
```diff
@@ -1,7 +1,7 @@
 import 'dart:async';
 
 import 'package:powersync_attachments_helper/powersync_attachments_helper.dart';
-import 'package:powersync/powersync.dart';
+import 'package:powersync_core/powersync_core.dart';
```
Comment (stevensJourney) [inline packages/powersync_attachments_helper/example/getting_started.dart:4]: Is this the recommended approach for this package? Would it not be imported from the `powersync` or SQLCipher version package?

File: packages/powersync_sqlcipher/example/main.dart [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+  if (!kIsWeb) {
+    final dir = await getApplicationSupportDirectory();
+    path = join(dir.path, 'powersync-dart.db');
+  }
+
+  // Setup the database.
+  final cipherFactory = PowerSyncSQLCipherOpenFactory(
+      path: path, key: "sqlcipher-encryption-key");
+
+  db = PowerSyncDatabase.withFactory(cipherFactory, schema: schema);
```
Comment (stevensJourney) [inline packages/powersync_sqlcipher/example/main.dart:40]: Is it possible to extend the regular `PowerSyncDatabase` constructor with an option for the cipher key? That might make the API a bit easier to use.

File: packages/powersync_attachments_helper/example/getting_started.dart (hunk 6 above)
Comment (rkistner) [reply packages/powersync_attachments_helper/example/getting_started.dart:4]: For a typical app you'd use one of those two. If you need to be compatible with both (like this helper lib), you'd use `powersync_core`.

Comment (mugikhan) [reply packages/powersync_attachments_helper/example/getting_started.dart:4]: I can add a comment showing that you can use `powersync` or `powersync_sqlcipher` in a client app?

Comment (stevensJourney) [reply packages/powersync_attachments_helper/example/getting_started.dart:4]: I see. I didn't scan through the entire file. I thought this example snippet actually instantiated a PowerSync client. I see it's just importing the type from the core package.

File: packages/powersync_sqlcipher/example/main.dart (hunk 7 above)
Comment (mugikhan) [reply packages/powersync_sqlcipher/example/main.dart:40]: As discussed, the current `PowerSyncDatabase` class only has factory constructors which doesn't allow it to be extended in Dart.

**Review by stevensJourney — DISMISSED** [review]

File: packages/powersync/README.md [hunk 8]
```diff
@@ -2,14 +2,20 @@
...
-This package (`powersync`) is the PowerSync client SDK for Dart/Flutter.
+This package (`powersync`) is the PowerSync client SDK for Flutter.
 
 See a summary of features [here](https://docs.powersync.com/client-sdk-references/flutter).
 
+> **Note**
+> This SDK is recommended for Flutter applications that do not require on-device encryption.
+
+> - For Flutter applications that require encryption, use the [`powersync_sqlcipher`](https://pub.dev/packages/powersync_sqlcipher) SDK.
+> - For use cases such as server-side Dart or non-Flutter Dart environments, use the [`powersync_core`](https://pub.dev/packages/powersync_core) SDK.
```
Comment (rkistner) [inline packages/powersync/README.md:17]:
```suggestion
> This package is recommended for Flutter applications that do not require on-device encryption.

> - For Flutter applications that require encryption, use the [`powersync_sqlcipher`](https://pub.dev/packages/powersync_sqlcipher) package.
> - For use cases such as server-side Dart or non-Flutter Dart environments, use the [`powersync_core`](https://pub.dev/packages/powersync_core) package.
```

File: packages/powersync_core/pubspec.yaml [hunk 9]
```diff
@@ -0,0 +1,43 @@
+name: powersync_core
+version: 0.0.1
```
Comment (rkistner) [inline packages/powersync_core/pubspec.yaml:2]: Can we start at a higher version, e.g. 1.0.0? While it's technically a new package, it was based on the powersync one, which is already stable.

File: packages/powersync_sqlcipher/pubspec.yaml [hunk 10]
```diff
@@ -0,0 +1,31 @@
+name: powersync_sqlcipher
+version: 0.0.1
```
Comment (rkistner) [inline packages/powersync_sqlcipher/pubspec.yaml:2]: Can we start at `0.1.0`? I generally view 0.0.x versions as "very experimental".

**Review by rkistner — COMMENTED** [review]
The melos.yml `version` -> `preCommit` command needs to be updated - the version.dart file has moved.

File: packages/powersync_core/pubspec.yaml (hunk 9 above)
Comment (mugikhan) [reply packages/powersync_core/pubspec.yaml:2]: Yes I will run melos version and bump the version to a stable 1.0.0, once any other issues are resolved.

File: packages/powersync_sqlcipher/pubspec.yaml (hunk 10 above)
Comment (mugikhan) [reply packages/powersync_sqlcipher/pubspec.yaml:2]: Will do with melos version once the issues are resolved.

**Comment by mugikhan** [comment]
> The melos.yml `version` -> `preCommit` command needs to be updated - the version.dart file has moved.

Should I move the version file back to the `powersync` package, it would live in the lib folder since it refers to that libraries version and not `powersync_core`?

**Review by rkistner — DISMISSED** [review]

**Review by rkistner — APPROVED** [review]

**Pull request merged**